### PR TITLE
Handle Lock type in VB

### PR DESCRIPTION
**Pull request by jjonescz**
Test plan: https://github.com/dotnet/roslyn/issues/71888

- Warns when a value of type `Lock` is used in `SyncLock` statement (users could expect it to behave like in C# but it currently does not).
- Warns on conversions of values of type `Lock` to a different type (users could expect the new locking to take effect on the target value which isn't supported in VB anyway but we cannot warn on the `SyncLock` statement without the type known to be `Lock` at compile time there).

### CODE REVIEW

File: src/Compilers/VisualBasic/Portable/VBResources.resx [hunk 1]
```diff
@@ -5695,4 +5695,16 @@
   <data name="ERR_InvalidExperimentalDiagID" xml:space="preserve">
     <value>The diagnosticId argument to the 'Experimental' attribute must be a valid identifier</value>
   </data>
+  <data name="WRN_LockTypeUnsupported" xml:space="preserve">
+    <value>A value of type 'System.Threading.Lock' in SyncLock will use likely unintended monitor-based locking. Consider manually calling 'Enter' and 'Exit' methods in a Try/Finally block instead.</value>
```
Comment (cston) [inline src/Compilers/VisualBasic/Portable/VBResources.resx:5699]: >'Enter' and 'Exit' [](http://example.com/codeflow?start=139&length=18)

Should this be "'EnterLockScope' and 'Dispose' methods" instead?

File: src/Compilers/VisualBasic/Portable/Binding/Binder_Conversions.vb [hunk 2]
```diff
@@ -1723,6 +1732,12 @@ DoneWithDiagnostics:
             End If
         End Sub
 
+        Private Sub WarnOnLockConversion(sourceType As TypeSymbol, syntax As SyntaxNode, diagnostics As BindingDiagnosticBag)
```
Comment (cston) [inline src/Compilers/VisualBasic/Portable/Binding/Binder_Conversions.vb:1735]: >Private [](http://example.com/codeflow?start=8&length=7)

`Shared`?

**Review by cston — APPROVED** [review]

File: src/Compilers/VisualBasic/Portable/VBResources.resx (hunk 1 above)
Comment (jjonescz) [reply src/Compilers/VisualBasic/Portable/VBResources.resx:5699]: No, the `Lock.Scope` type returned by `EnterLockScope` cannot be used in VB as it's a `ref struct`, so this warning suggests using `Lock.Enter` and `Lock.Exit` instead.

Comment (AlekseyTs) [inline src/Compilers/VisualBasic/Portable/VBResources.resx:5699]: >'Enter' and 'Exit' [](http://example.com/codeflow?start=139&length=18)

I do not see these APIs in the feature speclet. Where are they defined?

File: src/Compilers/VisualBasic/Portable/Symbols/TypeSymbolExtensions.vb [hunk 3]
```diff
@@ -1297,6 +1297,12 @@ Namespace Microsoft.CodeAnalysis.VisualBasic.Symbols
             Return typeSymbol.IsWellKnownCompilerServicesTopLevelType("IsExternalInit")
         End Function
 
+        <Extension>
+        Friend Function IsWellKnownTypeLock(typeSymbol As TypeSymbol) As Boolean
+            Return typeSymbol.Name = "Lock" AndAlso typeSymbol.ContainingType Is Nothing AndAlso
```
Comment (AlekseyTs) [inline src/Compilers/VisualBasic/Portable/Symbols/TypeSymbolExtensions.vb:1302]: >typeSymbol.Name = "Lock" [](http://example.com/codeflow?start=19&length=24)

Should arity be checked?

File: src/Compilers/VisualBasic/Test/Semantic/Binding/SyncLockTests.vb [hunk 4]
```diff
@@ -473,5 +474,194 @@ End Module
...
+    End Sub
+End Module
+
+Namespace System.Threading
+    Public Class Lock
+    End Class
+End Namespace
+"
+            CreateCompilation(source).AssertTheseDiagnostics(
+"BC42508: A value of type 'System.Threading.Lock' in SyncLock will use likely unintended monitor-based locking. Consider manually calling 'Enter' and 'Exit' methods in a Try/Finally block instead.
```
Comment (AlekseyTs) [inline src/Compilers/VisualBasic/Test/Semantic/Binding/SyncLockTests.vb:495]: >Consider manually calling 'Enter' and 'Exit' methods in a Try/Finally block instead. [](http://example.com/codeflow?start=112&length=84)

Consider adding a test verifying that following this suggestion is going to compile.

File: src/Compilers/VisualBasic/Test/Semantic/Binding/SyncLockTests.vb [hunk 5]
```diff
@@ -473,5 +474,194 @@ End Module
...
+        Public Sub LockType_CastToObject()
+            Dim source = "
+Imports System.Threading
+
+Module Program
+    Sub Main()
+        Dim l = New Lock()
+        Dim o As Object = l
+
+        o = DirectCast(l, Object)
```
Comment (AlekseyTs) [inline src/Compilers/VisualBasic/Test/Semantic/Binding/SyncLockTests.vb:511]: >DirectCast(l, Object) [](http://example.com/codeflow?start=12&length=21)

Please add tests with conversion to itself, i.e. `Lock` type

File: src/Compilers/VisualBasic/Portable/Symbols/TypeSymbolExtensions.vb [hunk 6]
```diff
@@ -1297,6 +1297,12 @@ Namespace Microsoft.CodeAnalysis.VisualBasic.Symbols
             Return typeSymbol.IsWellKnownCompilerServicesTopLevelType("IsExternalInit")
         End Function
 
+        <Extension>
+        Friend Function IsWellKnownTypeLock(typeSymbol As TypeSymbol) As Boolean
```
Comment (AlekseyTs) [inline src/Compilers/VisualBasic/Portable/Symbols/TypeSymbolExtensions.vb:1301]: It looks like this function doesn't have much test coverage, especially for negative results.

**Comment by AlekseyTs** [comment]
Done with review pass (commit 2)

File: src/Compilers/VisualBasic/Portable/VBResources.resx (hunk 1 above)
Comment (jjonescz) [reply src/Compilers/VisualBasic/Portable/VBResources.resx:5699]: In the runtime API proposal: https://github.com/dotnet/runtime/issues/34812

File: src/Compilers/VisualBasic/Test/Semantic/Binding/SyncLockTests.vb (hunk 4 above)
Comment (jjonescz) [reply src/Compilers/VisualBasic/Test/Semantic/Binding/SyncLockTests.vb:495]: It won't compile currently as the Lock type defined in those tests is just a stub. I can define the Enter/Exit methods but seems that such test would not provide much benefit. Better would be to have such test against real net9 runtime in the future.

File: src/Compilers/VisualBasic/Portable/VBResources.resx (hunk 1 above)
Comment (AlekseyTs) [reply src/Compilers/VisualBasic/Portable/VBResources.resx:5699]: > In the runtime API proposal: https://github.com/dotnet/runtime/issues/34812

I suggest updating the C# speclet accordingly then. Also, according to that link, it looks like the ```EnterLockScope``` is obsolete/changed, but we are still using old name.

Comment (jjonescz) [reply src/Compilers/VisualBasic/Portable/VBResources.resx:5699]: > I suggest updating the C# speclet accordingly then

Not sure what you mean, the C# feature doesn't use these Enter/Exit methods, so they probably shouldn't be part of the speclet?

> Also, according to that link, it looks like the `EnterLockScope` is obsolete/changed, but we are still using old name.

Yes, I know, waiting for the API to be approved and then I will change to the final name.

File: src/Compilers/VisualBasic/Portable/Binding/Binder_Conversions.vb [hunk 7]
```diff
@@ -1723,6 +1732,13 @@ DoneWithDiagnostics:
             End If
         End Sub
 
+        Private Shared Sub WarnOnLockConversion(sourceType As TypeSymbol, targetType As TypeSymbol, syntax As SyntaxNode, diagnostics As BindingDiagnosticBag)
+            If sourceType IsNot Nothing AndAlso sourceType.IsWellKnownTypeLock() AndAlso
+               targetType IsNot Nothing AndAlso Not targetType.IsWellKnownTypeLock() Then
```
Comment (AlekseyTs) [inline src/Compilers/VisualBasic/Portable/Binding/Binder_Conversions.vb:1737]: >targetType IsNot Nothing [](http://example.com/codeflow?start=15&length=24)

Can this check ever fail?

[the remark this answers is not in the document]
Comment (AlekseyTs) [reply src/Compilers/VisualBasic/Portable/Binding/Binder_Conversions.vb:1737]: Callers can check the kind before calling this helper

File: src/Compilers/VisualBasic/Portable/VBResources.resx (hunk 1 above)
Comment (AlekseyTs) [reply src/Compilers/VisualBasic/Portable/VBResources.resx:5699]: > Not sure what you mean, the C# feature doesn't use these Enter/Exit methods, so they probably shouldn't be part of the speclet?

Aren't these APIs part of the type. I do not see any harm in mentioning them. But I do see a benefit. At least we wouldn't have this conversation if the speclet had the information.

File: src/Compilers/VisualBasic/Test/Semantic/Binding/SyncLockTests.vb (hunk 4 above)
Comment (AlekseyTs) [reply src/Compilers/VisualBasic/Test/Semantic/Binding/SyncLockTests.vb:495]: > It won't compile currently as the Lock type defined in those tests is just a stub. I can define the Enter/Exit methods but seems that such test would not provide much benefit. Better would be to have such test against real net9 runtime in the future.

Please add a prototype comment then.

File: src/Compilers/VisualBasic/Test/Semantic/Binding/SyncLockTests.vb [hunk 8]
```diff
@@ -473,5 +474,328 @@ End Module
...
+Module Program
+    Sub Main()
+        Dim l = New Threading.Lock()
+        SyncLock l
+        End SyncLock
+    End Sub
+End Module
+
+Namespace Threading
+    Public Class Lock
```
Comment (AlekseyTs) [inline src/Compilers/VisualBasic/Test/Semantic/Binding/SyncLockTests.vb:553]: >Lock [](http://example.com/codeflow?start=17&length=4)

For completeness, consider adding a test for correct namespace, but "wrong" type name.

**Comment by AlekseyTs** [comment]
Done with review pass (commit 6)

File: src/Compilers/VisualBasic/Portable/VBResources.resx (hunk 1 above)
Comment (jjonescz) [reply src/Compilers/VisualBasic/Portable/VBResources.resx:5699]: Just adding them to the Lock type as mentioned in the speclet today would probably imply we check for existence of those methods and fail if they are missing. But I guess I can mention them separately. https://github.com/dotnet/csharplang/pull/7921

File: src/Compilers/VisualBasic/Test/Semantic/Binding/SyncLockTests.vb (hunk 4 above)
Comment (jjonescz) [reply src/Compilers/VisualBasic/Test/Semantic/Binding/SyncLockTests.vb:495]: > Please add a prototype comment then.

That would block merging the feature until we have net9, not sure we want that? I added a bullet point to the test plan, though.

File: src/Compilers/VisualBasic/Portable/VBResources.resx (hunk 1 above)
Comment (AlekseyTs) [reply src/Compilers/VisualBasic/Portable/VBResources.resx:5699]: > Just adding them to the Lock type as mentioned in the speclet today would probably imply we check for existence of those methods and fail if they are missing.

No would not. Definition of the type is just a definition of the type. There is no requirement to drop un-interesting members. The only motivation is to save some space in the document. But these members are interesting. Presence of which members and under which conditions is important for the feature is explicitly stated elsewhere in the speclet.

File: src/Compilers/VisualBasic/Test/Semantic/Binding/SyncLockTests.vb (hunk 4 above)
Comment (AlekseyTs) [reply src/Compilers/VisualBasic/Test/Semantic/Binding/SyncLockTests.vb:495]: > That would block merging the feature until we have net9, not sure we want that? 

That just means that we will have to open an issue to track completion of the work if we are fine with merging with the gap. The same decision will have to be made for items in the Test Plan. So, in that sense they are all blocking the merge, unless an explicit decision is made to not block on the specific item.

**Review by AlekseyTs — APPROVED** [review]
LGTM (commit 8)

**Comment by jjonescz** [comment]
@cston for another look as I changed the implementation slightly. Thanks.

**Review by cston — APPROVED** [review]

**Pull request merged**